### PR TITLE
[DataGrid] Save paging state in URL

### LINKED ISSUE #2839: feat: add url support to DataGrid
# 🙋 Feature Request

We have a grid with a link to a detail page, if we click the link and then return to the page with the grid, the paging info is lost. It would be great if we can automatically store that information in the url.

## 🤔 Expected Behavior

Change the url if the user is paging in the grid, by setting the url through the navigationmanager.

## 😯 Current Behavior

Paging information is lost

## 💁 Possible Solution

You tell the GridItemsProvider to load the current settings from the url upon load and save it to the url when changed. Possibly with a boolean to turn on this behavior.

## 🔦 Context

I considered building this myself, then thought I cannot possibly be to only one having this challenge.

## 💻 Examples

1. Page `/users` has a grid with (server-side) paging
2. The user goes to page 2 with the pagination control
3. Automatically the url is changed to `/users?page=2`
4. the user goes to a detail page
5. the user presses the back button in the browser
6. the user browses to `/users?page=2`
7. the GridItemsProvider loads the paging information from the url
8. The grid loads the correct data (page 2)

### PR DESCRIPTION
**Pull request by svrooij**
Fixed #2839

# Pull Request

## 📖 Description

Paging or sorting for a data grid will now save the state in the url, if the developer opt-ed in for this on that grid. It also supports a prefix, in case you would have multiple grids on a single page.

### 🎫 Issues

* #2839 

### General

- [ ] I have added tests for my changes.
- [x] I have tested my changes.
- [ ] I have updated the project documentation to reflect my changes.
- [ ] I have read the [CONTRIBUTING](https://github.com/microsoft/fluentui-blazor/blob/master/docs/contributing.md) documentation and followed the [standards](https://www.fast.design/docs/community/code-of-conduct/#our-standards) for this project.

### Component-specific

- [ ] I have added a new component
- [ ] I have added [Unit Tests](https://github.com/Microsoft/fluentui-blazor/blob/master/unit-tests.md) for my new component
- [X] I have modified an existing component
- [X] I have validated the [Unit Tests](https://github.com/Microsoft/fluentui-blazor/blob/master/unit-tests.md) for an existing component `404`

### CODE REVIEW

**Review by vnbaaij — APPROVED** [review]

File: src/Core/Components/DataGrid/FluentDataGrid.razor.cs [hunk 1]
```diff
@@ -245,6 +248,18 @@ public partial class FluentDataGrid<TGridItem> : FluentComponentBase, IHandleEve
...
+    /// Gets or sets a value indicating whether the grid should save its paging state in the URL.
+    /// </summary>
+    [Parameter]
+    public bool SaveStateInUrl { get; set; }
+
+    /// <summary>
+    /// Gets or sets a prefix to use when saving the grid state in the URL.
+    /// </summary>
+    /// <remarks>Only relevant when <see cref="SaveStateInUrl"/> is set to <see langword="true"/> on multiple grids on a single page.</remarks>
+    public string? SaveStatePrefix { get; set; }
```
Comment (dvoituron) [inline src/Core/Components/DataGrid/FluentDataGrid.razor.cs:261]: Did you miss the `Parameter` attribute?

File: src/Core/Components/DataGrid/FluentDataGrid.razor.cs [hunk 2]
```diff
@@ -795,6 +814,52 @@ private void CloseColumnResize()
...
+            if (query.AllKeys.Contains($"{SaveStatePrefix}top") && int.TryParse(query[$"{SaveStatePrefix}top"]!, out int itemsPerPage))
+            {
+                Pagination.ItemsPerPage = itemsPerPage;
+            }
+        }
+    }
+
+    private void SaveStateToQueryString()
+    {
+        if (SaveStateInUrl)
```
Comment (dvoituron) [inline src/Core/Components/DataGrid/FluentDataGrid.razor.cs:849]: For a better reading of the code, can you replace this with

```csharp
if (SaveStateInUrl)
{
    return;
}
```

File: src/Core/Components/DataGrid/FluentDataGrid.razor.cs [hunk 3]
```diff
@@ -795,6 +814,52 @@ private void CloseColumnResize()
         StateHasChanged();
     }
 
+    private void LoadStateFromQueryString(string queryString)
+    {
```
Comment (dvoituron) [inline src/Core/Components/DataGrid/FluentDataGrid.razor.cs:818]: Can you add this test, like you did for the Save method (even if the test is already present in the `OnInitialize`, which remains interesting for performance questions)?

```csharp
if (SaveStateInUrl)
{
    return;
}
```

**Review by dvoituron — CHANGES_REQUESTED** [review]

File: src/Core/Components/DataGrid/FluentDataGrid.razor.cs (hunk 1 above)
Comment (svrooij) [reply src/Core/Components/DataGrid/FluentDataGrid.razor.cs:261]: Yes 🤭

**Review by dvoituron — APPROVED** [review]

**Comment by svrooij** [comment]
@vnbaaij and @dvoituron the tests work on my machine, I see the build is also failing on the dev branch. Is that something that needs investigating?

For one of my projects I use the [GitHub Actions test logger](https://www.nuget.org/packages/GitHubActionsTestLogger) to produce these [nice looking stats](https://github.com/svrooij/WingetIntune/actions/runs/11814350986#summary-32913235379) in the action. It will also show exactly which test failed (if any). Is that something you might also want in your [pipeline](https://github.com/svrooij/WingetIntune/blob/81a4bb0e887946bb420f850f996c61ab06dee423/.github/workflows/build.yml#L62-L63)?

**Comment by dvoituron** [comment]
> @vnbaaij and @dvoituron the tests work on my machine, I see the build is also failing on the dev branch. Is that something that needs investigating?

We have a similar process in our Pipeline:
{0C01C70E-8C4D-4805-9556-5637A2C960C6}

But the problem comes with user rights (for automatic publication in the RP comment stream): Repo owners see this report but not you :-)

But in this case, it's more a .NET9 issue. @vnbaaij did you changed something?

**Comment by svrooij** [comment]
One log line said it could not download dotnet 9.0 ga, so I think it has to do with the official release of dotnet9

**Comment by vnbaaij** [comment]
> One log line said it could not download dotnet 9.0 ga, so I think it has to do with the official release of dotnet9

There was a reference still to .net9 preview in a pipeline. This has been corrected



> But in this case, it's more a .NET9 issue. @vnbaaij did you changed something?

Yes, but this test has already been fixed

**Pull request merged**